src/ser.rs

```diff
@@ -0,0 +1,183 @@
+        self.update_parents(eclass, &enode);
+        EGraph::<L, ()>::update_memo_from_parent(&mut self.memo, &enode, &eclass);
+        let class = self.classes[usize::from(eclass)].as_mut().unwrap();
+        class.nodes.push(enode);
+        return class.nodes.last().unwrap();
+    }
+
+    fn update_parents(&mut self, parent: Id, enode: &L) {
+        enode.children().iter().for_each(|u| self.classes[usize::from(*u)].as_mut().unwrap()
+            .parents.push((enode.clone(), parent)));
+    }
+
+    fn line_to_tuple(line: &str) -> (String, Vec<Id>) {
+        let v: Vec<_> = line.split(' ').filter(|s| s.len() > 0).collect();
+        (v[0].into(), v[1..].iter().map(
+            |u| Id::from(u.parse::<usize>().unwrap())).collect())
+    }
+}
+
+impl Deserialization for EGraph<SymbolLang, ()> {
+
+    fn from_tuples<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<Id>)>) -> (Self, ColorPalette) {
+        let mut g = EGraph::<SymbolLang, ()>::default();
+        let mut leaf_ops = Vec::new();
+        let mut color_unions = Vec::new();
+        for (op, vertices) in tuples {
+            let op = op.to_string();
+            if op == "?~" { color_unions.push(vertices.clone()); continue; }
+            for id in vertices { g.add_class(*id); }
+            let target = vertices[0];
+            let sources = vertices[1..].to_vec();
+            let enode = g.add_node(target, SymbolLang::new(op, sources));
+            if enode.is_leaf() { leaf_ops.push(enode.op_id()) }
+        }
+        //g.rebuild_classes();
+        g.rebuild();
+        for op in leaf_ops {
+            let eclass = g.classes_by_op.get(&op).unwrap();
+            let min = *eclass.iter().min().unwrap();
+            let others = eclass.iter().filter_map(|u| { if *u != min { Some(*u) } else { None } })
+                               .collect::<Vec<Id>>();
+            others.iter().for_each(|u| { g.union(min, *u); g.rebuild(); });
+        }
+        g.rebuild();
+        // Performed deferred color merges
+        let mut palette = ColorPalette::new();
+        for vertices in color_unions {
+            let v0 = g.find(vertices[0]);
+            let c = palette.get(&mut g, v0);
+            let u = vertices[1];
+            for v in &vertices[2..] { g.colored_union(c, u, *v); }
+        }
+        g.rebuild();
+        (g, palette)
+    }
+
+    fn from_tuples_int<'a>(tuples: impl Iterator<Item=&'a (impl ToString + 'a, Vec<usize>)>) -> (Self, ColorPalette) {
+        EGraph::<SymbolLang, ()>::from_tuples(
+            tuples.map(|(op, members)|
+                (op.to_string(), members.iter().map(|v| Id::from(*v)).collect::<Vec<Id>>()))
+                .collect::<Vec<_>>().iter())
+    }
+
+    fn from_tuples_text(in_: &mut impl Read) -> Result<(Self, ColorPalette)> {
+        Ok(EGraph::<SymbolLang, ()>::from_tuples(BufReader::new(in_).lines()
+            .map(|r| r.unwrap().trim().to_string())
+            .filter(|s| s.len() > 0)
```

Comment:
What about tests? I would expect a test that tries to serialize after some basic rewriting and the desrialize and compare